feat : 카카오페이 API 를 이용한 포인트 충전기능 구현

romay99 opened the pull request:
**TO-BE**
카카오페이 API 를 이용한 포인트 충전기능 구현 완료했습니다.

충전이 성공하면 내부DB에 거래기록이 저장됩니다.

KakaoPayController -> prepareChargePoint() 메서드가 호출된 뒤 사용자 인증되면 approvedPointCharge() 메서드로 redirect 됨.

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [x] API 테스트

romay99 committed:
feat : 카카오페이 결제 구현중

romay99 committed:
feat : 카카오페이 결제 기능 구현 완료
사용자 계정에 포인트추가, DB에 기록남기기
2가지 구현 필요함

romay99 committed:
Merge branch 'develop' into feat/recharge_points_kakao

romay99 committed:
feat : 카카오페이 API 로 포인트 충전기능 구현
포인트 충전을 완료하면, 내부 거래기록 DB에 저장

focandlol commented on src/main/java/com/watchtogether/watchtogether/api/service/KakaoPayApiService.java:
```diff
@@ -0,0 +1,129 @@
+    return response;
+  }
+
+  @Transactional
+  public KakaoPayApproveResponseDto approvedPointCharge(String pgToken, String memberId,
+      int amount) {
+    // 요청보내는 Uri 생성
```
api 호출과 db저장 부분이 @Transactional로 묶여 있는 것 같습니다.
api 호출하는 부분, db와 커넥션이 있는 부분을 각각 따로 메서드로 만들어서 db와 커넥션이 있는 부분에만 @Transactional을 붙이는 것도 괜찮을 것 같습니다:)

```
public KakaoPayApproveResponseDto apiRequest(String pqToken, String memberId){
  //api 호출하는 코드
  return response.body();
}

@Transactional
public void save(String memberId, int amount){
    Member member = memberRepository.findByMemberId(memberId).get();

    transactionHistoryService.usePoint(member, amount, TransactionDetail.CHARGE);
}
```
이렇게 하고 컨트롤러에서 따로 호출하거나 아예 클래스를 나눠도 좋을 것 같습니다.

romay99 replied:
아래에 DB에 저장하는 부분을 메서드로 빼내어 컨트롤러에서 호출하게끔 수정하였습니다. 좋은의견 감사드립니다!

( 컨트롤러 -> API 호출 -> return 받은 데이터로 다시 DB 저장하는 메서드 호출 -> DB 저장)

KorMH commented:
카카오페이 결제 구현하는 기능에 대한 코드를 잘 보고 갑니다.

mingming-mentor commented on src/main/java/com/watchtogether/watchtogether/api/dto/KakaoPayApproveDto.java:
```diff
@@ -0,0 +1,18 @@
+
+  private String cid;
+  private String tid;
+  private String partner_order_id;
+  private String partner_user_id;
+  private String pg_token;
```
카멜케이스로 작성 후 @JsonProperty를 이용해서 맵핑해주시면 좋을 것 같아요. 기본적으로 자바에서는 카멜케이스로 변수명을 정하는 것이 기본 컨벤션이라 참고해보시면 좋을 것 같습니다

romay99 replied:
네 매니저님 변수명은 카멜케이스로 변경후 @JsonProperty 로 매핑해주었습니다!

mingming-mentor commented on src/main/java/com/watchtogether/watchtogether/api/service/KakaoPayApiService.java:
```diff
@@ -0,0 +1,129 @@
+  private final MemberRepository memberRepository;
+
+  // API KEY 를 application.properties 에서 가져온다.
+  public KakaoPayApiService(@Value("${kakaoPay.apiKey}") String key,
+      @Value("${kakaoPay.cidKey}") String cidKey, RestTemplateBuilder builder,
+      KakaoPayRedisService kakaoPayRedisService, TransactionHistoryService service
+      , MemberRepository memberRepository) {
+    this.API_KEY = key;
+    this.CID_KEY = cidKey;
+    this.restTemplate = builder.build();
+    this.kakaoPayRedisService = kakaoPayRedisService;
+    this.transactionHistoryService = service;
+    this.memberRepository = memberRepository;
+  }
```
@RequiredArgsConstrutor를 쓰시고 API_KEY와 CID_KEY는 변수위에 @Value를 쓰셔서 바로 주입받아도 괜찮을 것 같아요. restTemplate 부분만 @PostConstructor를 이용해 build해주셔도 좋구요.
이 부분은 취향차이이므로 그냥 이렇게 하는 방법도 있다 정도의 코멘트로 봐주시면 좋겠습니다.

romay99 replied:
restTemplate 은 별도로 Bean 을 생성해 DI 받는것으로 수정하여 @RequiredArgsConstrutor 를 사용하는 쪽으로 수정하였습니다!

mingming-mentor reviewed (approved):
수고 많으셨어요

romay99 committed:
refactor : KakaoPayApiService API 호출, DB 저장 코드 분리
KakaoPayApiService API 호출, DB 저장코드 분리
KakaoPayApproveDto 내부 카멜케이스로 변경
DI 주입받는 방식 변경

romay99 committed:
refactor : KakaoPay 관련 DTO들 전부 카멜케이스로 변경

romay99 committed:
Merge branch 'develop' of https://github.com/romay99/WatchTogether into feat/recharge_points_kakao

# Conflicts:
#	src/main/java/com/watchtogether/watchtogether/config/SecurityConfig.java

romay99 committed:
fix : develop 브런치와 merge
